### PR TITLE
TcpClient.ConnectAsync(EndPoint)

### LINKED ISSUE #43111: Missing `System.Net.Sockets.TcpClient.Connect(IPEndPoint remoteEP)` overload.
## Background and Motivation

This issue proposes to add `System.Net.Sockets.TcpClient.ConnectAsync(IPEndPoint remoteEP)` method. 

This issue was filed based on #40750 and comment https://github.com/dotnet/runtime/pull/40750#issuecomment-704112931 by  @geoffkizer.

## Proposed API

```diff
namespace System.Net.Sockets
public class TcpClient : IDisposable
{

    public void Connect(IPAddress address, int port);
    public void Connect(IPAddress[] ipAddresses, int port);
    public void Connect(IPEndPoint remoteEP);
    public void Connect(string hostname, int port);

    public Task ConnectAsync(IPAddress address, int port);
    public Task ConnectAsync(IPAddress[] addresses, int port);
+    public Task ConnectAsync(IPEndPoint remoteEP);
    public Task ConnectAsync(string host, int port);

    public ValueTask ConnectAsync (IPAddress address, int port, CancellationToken cancellationToken);
    public ValueTask ConnectAsync (IPAddress[] addresses, int port, CancellationToken cancellationToken);
+    public ValueTask ConnectAsync(IPEndPoint remoteEP,  CancellationToken cancellationToken);
    public ValueTask ConnectAsync (string host, int port, CancellationToken cancellationToken);
}
```

## Usage Examples

Usage example is the same as for `System.Net.Sockets.TcpClient.Connect(IPEndPoint remoteEP)`.

## Alternative Designs

N/A.

## Risks

N/A

### PR DESCRIPTION
**Pull request by MartyIX**
Tentative attempt to fix #43111.

I'm not familiar with this code so it may be totally off. I'm not sure. Feedback is welcome. Or if somebody wants to finish the job, I'm not against.

Friendly ping @karelz @geoffkizer

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=44110) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=44110) before we can accept your contribution.<br/><br/>:x: MartyIX [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=44110)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=44110) it.</sub>

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/TcpClientTest.cs [hunk 1]
```diff
@@ -168,6 +169,10 @@ public async Task ConnectAsync_DnsEndPoint_Success(int mode)
                         await client.ConnectAsync(addresses[0], port, CancellationToken.None);
                         break;
                     case 8:
+                        addresses = await Dns.GetHostAddressesAsync(host);
+                        await client.ConnectAsync(new IPEndPoint(addresses[0], port), CancellationToken.None);
+                        break;
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/TcpClientTest.cs:174]: Can we add a test for the other overload being added, too?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs [hunk 2]
```diff
@@ -283,6 +286,9 @@ private async Task CompleteConnectAsync(Task task)
         public ValueTask ConnectAsync(IPAddress[] addresses, int port, CancellationToken cancellationToken) =>
             CompleteConnectAsync(Client.ConnectAsync(addresses, port, cancellationToken));
 
+        public ValueTask ConnectAsync(EndPoint remoteEP, CancellationToken cancellationToken) =>
+            CompleteConnectAsync(Client.ConnectAsync(remoteEP, cancellationToken));
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:290]: I know it's inconsistent with the rest of the file, but we're moving to a model of using XML comments for documentation.  Can you add XML comments for the two new methods?  You can copy liberally from the existing docs (https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.tcpclient.connectasync?view=net-5.0).

**Review by stephentoub — COMMENTED** [review]
A few comments, but otherwise LGTM. Thanks.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs (hunk 2 above)
Comment (MartyIX) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:301]: I added a comment. Hopefully, it's OK.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/TcpClientTest.cs (hunk 1 above)
Comment (MartyIX) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/TcpClientTest.cs:175]: I added a test. Not sure whether those CI errors are caused by the test though. Kind of seems unrelated.

**Review by antonfirsov — APPROVED** [review]
LGTM

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs [hunk 3]
```diff
@@ -610,9 +610,11 @@ public partial class TcpClient : System.IDisposable
         public System.Threading.Tasks.Task ConnectAsync(System.Net.IPAddress address, int port) { throw null; }
         public System.Threading.Tasks.Task ConnectAsync(System.Net.IPAddress[] addresses, int port) { throw null; }
         public System.Threading.Tasks.Task ConnectAsync(string host, int port) { throw null; }
+        public System.Threading.Tasks.Task ConnectAsync(System.Net.EndPoint remoteEP) { throw null; }
```
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs:613]: This should be IPEndPoint, *not* EndPoint. Similarly below.

**Comment by geoffkizer** [comment]
We really should have cancellation tests for the cancellation overloads, but since we don't today it's fine to skip them. Since it's literally just a pass-through to Socket it's not a big deal.

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs:613]: Good catch.  "Tcp"Client, after all.

Comment (MartyIX) [reply src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs]: Right.

**Comment by geoffkizer** [comment]
Thanks @MartyIX!

**Pull request merged**